### PR TITLE
FreeBSD: Fix includes to co-exist with dtrace's legacy opensolaris compat code 

### PR DESCRIPTION
**Pull request by mattmacy**
Fix header conflicts when building zfs with openzfs as a vendor import.

Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/Makefile.bsd [hunk 1]
```diff
@@ -25,6 +25,7 @@ CFLAGS+= -I${INCDIR}/os/freebsd
 CFLAGS+= -I${INCDIR}/os/freebsd/spl
 CFLAGS+= -I${INCDIR}/os/freebsd/zfs
 CFLAGS+= -include ${INCDIR}/os/freebsd/spl/sys/ccompile.h
+CFLAGS+= -I${INCDIR}/..
```
Comment (nivedita76) [inline module/Makefile.bsd:28]: Would that be cleaner as -I${SRCDIR}?

Comment (nivedita76) [reply module/Makefile.bsd:28]: Actually after my PR https://github.com/openzfs/zfs/pull/10493 you'd want -I${.OBJDIR} instead.

Comment (nivedita76) [reply module/Makefile.bsd:28]: Also, on linux we actually pull it into every compilation via `-include`. Should freebsd do the same?

Another thing here is that ${INCDIR}/spl no longer exists, should be removed from CFLAGS.

**Comment by nivedita76** [comment]
What is the case that's being solved? Is it with module/ being copied out of the openzfs tree?

Thanks.

**Comment by freqlabs** [comment]
Let's see how this looks after #10493 merges.

> What is the case that's being solved? Is it with module/ being copied out of the openzfs tree?

This is for the whole tree being vendored in FreeBSD's base src/sys/contrib/.

**Comment by nivedita76** [comment]
> Let's see how this looks after rebasing now that #10493 has merged.
> 
> > What is the case that's being solved? Is it with module/ being copied out of the openzfs tree?
> 
> This is for the whole tree being vendored in FreeBSD's base src/sys/contrib/.

What is the issue that causes? Would the tree structure be: src/sys/contrib/openzfs/{zfs_config.h, include/, module/, lib/,...} etc, or something else?

**Comment by mattmacy** [comment]
> > Let's see how this looks after rebasing now that #10493 has merged.
> > > What is the case that's being solved? Is it with module/ being copied out of the openzfs tree?
> > 
> > 
> > This is for the whole tree being vendored in FreeBSD's base src/sys/contrib/.
> 
> What is the issue that causes? Would the tree structure be: src/sys/contrib/openzfs/{zfs_config.h, include/, module/, lib/,...} etc, or something else?

Yes.

**Comment by mattmacy** [comment]
https://reviews.freebsd.org/D25477

**Comment by nivedita76** [comment]
> > > Let's see how this looks after rebasing now that #10493 has merged.
> > > > What is the case that's being solved? Is it with module/ being copied out of the openzfs tree?
> > > 
> > > 
> > > This is for the whole tree being vendored in FreeBSD's base src/sys/contrib/.
> > 
> > 
> > What is the issue that causes? Would the tree structure be: src/sys/contrib/openzfs/{zfs_config.h, include/, module/, lib/,...} etc, or something else?
> 
> Yes.

So why is the zfs_config.h change needed if it's in the same location relative to module/ -- what breaks?

I think changing the `#include` will break the out-of-tree building enabled by #10493 unless you add `${.OBJDIR:H}` in addition to `${INCDIR}/..` to the list of include directories.

**Comment by mattmacy** [comment]
> > > > Let's see how this looks after rebasing now that #10493 has merged.
> > > > > What is the case that's being solved? Is it with module/ being copied out of the openzfs tree?
> > > > 
> > > > 
> > > > This is for the whole tree being vendored in FreeBSD's base src/sys/contrib/.
> > > 
> > > 
> > > What is the issue that causes? Would the tree structure be: src/sys/contrib/openzfs/{zfs_config.h, include/, module/, lib/,...} etc, or something else?
> > 
> > 
> > Yes.
> 
> So why is the zfs_config.h change needed if it's in the same location relative to module/ -- what breaks?
> 
> I think changing the `#include` will break the out-of-tree building enabled by #10493 unless you add `${.OBJDIR:H}` in addition to `${INCDIR}/..` to the list of include directories.

It doesn't break out of tree building, but you're right that it's no longer needed. A bigger issue is that all the test fail for some reason. Tracking that down.

File: module/os/freebsd/zfs/sysctl_os.c [hunk 2]
```diff
@@ -124,10 +123,11 @@ SYSCTL_NODE(_vfs_zfs_vdev, OID_AUTO, cache, CTLFLAG_RW, 0, "ZFS VDEV Cache");
 SYSCTL_NODE(_vfs_zfs_vdev, OID_AUTO, mirror, CTLFLAG_RD, 0,
     "ZFS VDEV mirror");
 
+#ifdef ZFS_META_VERSION
```
Comment (freqlabs) [inline module/os/freebsd/zfs/sysctl_os.c:126]: Hmm well we do need to pull in this definition somehow.

Comment (freqlabs) [reply module/os/freebsd/zfs/sysctl_os.c:126]: Nevermind, that was moved to Makefile.bsd

File: include/os/freebsd/spl/sys/ccompile.h [hunk 3]
```diff
@@ -179,6 +181,9 @@ typedef	void zfs_kernel_param_t;
 #define	MUTEX_NOLOCKDEP 0
 #define	RW_NOLOCKDEP 0
 
+#if !defined(LOCORE) && !defined(_ASM)
```
Comment (behlendorf) [inline include/os/freebsd/spl/sys/ccompile.h:184]: Do I understand correctly that the `#if !defined(LOCORE) && !defined(_ASM)` check is used to detect if this code is not part of a vendor tree?  Assuming that's the case how does this differ from `BUILDING_ZFS` which is defined in the `Makefile.bsd`?  A comment for both of these cases would be helpful.

Comment (mattmacy) [reply include/os/freebsd/spl/sys/ccompile.h:184]: It's used because it's getting pulled in to assembly compilation

**Pull request merged**